Add Kubernetes 5.0 migration guide (#4933)
gravitational/teleport: docs/5.0/kubernetes-5.0-migration.md

---
title: Kubernetes 5.0 Migration Guide
description: How to migrate pre-5.0 Teleport clusters to new configuration
---

> # Kubernetes intergartion in 5.0

Reviewer: ```suggestion
# Kubernetes integration in 5.0
```

In release 5.0, Teleport has changed the [Kubernetes
integration](kubernetes-ssh.md) to improve configuration and user experience.
These changes are backwards compatible and there is no _required_ migration.
However, to get the most out of Teleport, you should consider migrating as
described below.

The main changes in 5.0 are:
- new `kubernetes_service` configuration section in `teleport.yaml`, decoupled
  from `proxy_service`
- support for multiple Kubernetes clusters per Teleport cluster (replaces the

>   need for [Trusted Clusters](admin-guide.md#trusted-clusters))

Reviewer: We should be a bit clearer that trusted clusters are still a useful feature for Teleport in general, just that you don't _need_ to use them to access multiple Kubernetes clusters any more.

```suggestion
  need to use [Trusted Clusters](admin-guide.md#trusted-clusters) to achieve this)
```

    enabled: yes
    listen_addr: 0.0.0.0:3026
    kubeconfig_file: /path/to/kubeconfig
```

After 5.0:

```yaml
# teleport.yaml
proxy_service:
  kube_listen_addr: 0.0.0.0:3026

kubernetes_service:
  enabled: yes
  listen_addr: 0.0.0.0:3027
  kubeconfig_file: /path/to/kubeconfig
```

The `listen_port` for `kubernetes_service` only needs to be exposed to the
Proxies within your cluster. This port will not be directly used by the
clients.

## Proxy in a Kubernetes pod of a single cluster

In this scenario, we're running a Teleport proxy inside of the Kubernetes
cluster and connecting to it using the pod service account credentials.

Before 5.0:

```yaml
# teleport.yaml
proxy_service:
  kubernetes:
    enabled: yes
    listen_addr: 0.0.0.0:3026
```

After 5.0:

```yaml
# teleport.yaml
proxy_service:
  kube_listen_addr: 0.0.0.0:3026

kubernetes_service:
  enabled: yes
  listen_addr: 0.0.0.0:3027
```

## Multiple Kubernetes clusters

In this scenario, we're running multiple Trusted Teleport clusters, one for
each Kubernetes cluster.

### Option 1: Teleport Kubernetes agents in each Kubernetes cluster

This is the recommended approach for handling multiple Kubernetes clusters.

> Instead of running multiple Teleport clusters (auth and proxy for each
> Kubernetes cluster), we will connect multiple `kubernetes_service` agents to
> the same root Teleport cluster.

Reviewer: Maybe Add some extra clarity about the fact that the connections are tunnelled?

```suggestion
Instead of running multiple Teleport clusters (auth and proxy for each
Kubernetes cluster), we will connect multiple `kubernetes_service` agents to
the same root Teleport cluster.

Each `kubernetes_service` will create its own reverse tunnel back to a Teleport
proxy to join the cluster.
```

---
# kubeconfig-all.yaml
apiVersion: v1
kind: Config
clusters:
- cluster:
    certificate-authority-data: ...
    server: https://10.0.0.1
  name: kube-1-server
- cluster:
    certificate-authority-data: ...
    server: https://10.0.0.2
  name: kube-2-server
contexts:
- context:
    cluster: kube-1-server
    user: kube-1-creds
  name: kube-1
- context:
    cluster: kube-2-server
    user: kube-2-creds
  name: kube-2
current-context: kube-1
users:
- name: kube-1-creds
  user:
    ...cluster-specific credentials...
- name: kube-2-creds
  user:
    ...cluster-specific credentials...
```

You can view the list of Kubernetes clusters using `tsh kube ls` and switch
between them using `tsh kube login <kube_cluster_name>`:

```sh
$ tsh kube ls
Kube Cluster Name Selected
----------------- --------
kube-1            *
kube-2

$ tsh kube login kube-2
Logged into kubernetes cluster "kube-2"
```

### Option 3: Keep using Trusted Clusters

You may keep using the Trusted Clusters as before and migrate them internally
to `kubernetes_service`, as described in the first two scenarios.

Note that `tsh kube ls` will **not** show all Kubernetes clusters across all

> Trusted Clusters. It only shows the Kubernetes clusters from the Teleort

Reviewer: ```suggestion
Trusted Clusters. It only shows the Kubernetes clusters from the Teleport
```

cluster you are currently logged into.

# RBAC

!!! note

>     Kubernetes RBAC integration only works with kubernetes_service. Make sure

Reviewer: ```suggestion
    Kubernetes RBAC integration only works with `kubernetes_service`. Make sure
```

    kubernetes_labels:
      # a user can only connect to clusters marked with 'test' label:
      'environment': 'test'
      # the wildcard ('*') means "any cluster"
      '*': '*'
      # labels can be specified as a list:
      'environment': ['test', 'staging']
      # regular expressions are also supported, for example the equivalent
      # of the list example above can be expressed as:
      'environment': '^test|staging$'
  # You can also strictly restrict access using deny kubernetes_labels.
  deny: {}
```

Next, define Kubernetes labels on the cluster side:

```yaml
# teleport.yaml
kubernetes_service:
  enabled: yes
  listen_addr: 0.0.0.0:3027
  kube_cluster_name: staging
  # Note: if you use kubeconfig_file, the below labels apply to all Kubernetes
  # clusters in that kubeconfig. To get different sets of labels, you need to
  # run separate kubernetes_service instances.
  labels:
    environment: staging
  commands:
  - name: project
    command: [fetch-project-name.sh]
    period: 1h
```

Users with a role that matches a specific Kubernetes cluster labels will be
able to see and connect to that cluster. Other users will neither see nor

> connect to it.

Reviewer: ```suggestion
be able to connect to it.
```

---
title: Kubernetes 5.0 Migration Guide
description: How to migrate pre-5.0 Teleport clusters to new configuration
---

> # Kubernetes integration in 5.0

Reviewer 2: ```suggestion
# Migrating to Teleport Kubernetes Access 5.0
```

cluster and connecting to it using a `kubeconfig_file`.

Before 5.0:

```yaml
# teleport.yaml
proxy_service:
  kubernetes:
    enabled: yes
    listen_addr: 0.0.0.0:3026
    kubeconfig_file: /path/to/kubeconfig
```

After 5.0:

```yaml
# teleport.yaml
proxy_service:
  kube_listen_addr: 0.0.0.0:3026

kubernetes_service:
  enabled: yes
  listen_addr: 0.0.0.0:3027
  kubeconfig_file: /path/to/kubeconfig
```

The `listen_port` for `kubernetes_service` only needs to be exposed to the
Proxies within your cluster. This port will not be directly used by the
clients.

## Proxy in a Kubernetes pod of a single cluster

In this scenario, we're running a Teleport proxy inside of the Kubernetes
cluster and connecting to it using the pod service account credentials.

Before 5.0:

```yaml
# teleport.yaml
proxy_service:
  kubernetes:
    enabled: yes
    listen_addr: 0.0.0.0:3026
```

After 5.0:

```yaml
# teleport.yaml
proxy_service:
  kube_listen_addr: 0.0.0.0:3026

kubernetes_service:
  enabled: yes
  listen_addr: 0.0.0.0:3027
```

## Multiple Kubernetes clusters

In this scenario, we're running multiple Trusted Teleport clusters, one for
each Kubernetes cluster.

> ### Option 1: Teleport Kubernetes agents in each Kubernetes cluster

Reviewer 2: ```suggestion
### Recommend Option 1: Teleport Kubernetes agents in each Kubernetes cluster
```
